src/main/java/com/squareup/javapoet/ClassName.java

```diff
@@ -122,7 +122,7 @@ public static ClassName bestGuess(String classNameString) {
 
     // Add the package name, like "java.util.concurrent", or "" for no package.
     int p = 0;
-    while (p < classNameString.length() && Util.isLowerCase(classNameString.charAt(p))) {
+    while (p < classNameString.length() && Character.isLowerCase(classNameString.charAt(p))) {
```

Comment:
A quick example https://ideone.com/68XtFO :)
IMHO class loader can be not always guilty. A filesystem might have troubles with such file names, a bad collation might be set in a database, HTTP header with charset might be not respected...
All this is getting better with time, but still it happens.